File: loopviewpagerlib/src/main/java/mocaris/com/loopviewpagerlib/LoopViewPagerAdapter2.java
package mocaris.com.loopviewpagerlib;

import android.content.Context;
import android.support.v4.view.PagerAdapter;
import android.support.v4.view.ViewPager;

import java.util.ArrayList;
import java.util.List;

/**
 * 无限轮播 viewPager
 */
public abstract class LoopViewPagerAdapter2<T> extends PagerAdapter implements ViewPager.OnPageChangeListener {

    protected Context context;
    protected ViewPager viewPager;
    protected List<T> mDatas;
    private List<T> provideList = new ArrayList<>();

    public LoopViewPagerAdapter2(Context context, ViewPager viewPager) {
        this.context = context;
        this.viewPager = viewPager;
        viewPager.addOnPageChangeListener(this);
    }

    public List<T> getDatas() {
        return mDatas;
    }

    public void setDatas(List<T> datas) {
        this.mDatas = datas;
        if (!provideList.isEmpty()) {
            provideList.clear();
        }
        provideList.addAll(datas);
        provideList.add(provideList.size() - 1, datas.get(0));//添加第一个
        provideList.add(0, datas.get(datas.size() - 1));//添加最后一个
        notifyDataSetChanged();
    }

    @Override
    public int getCount() {
        return provideList.size();
    }

    public T getItem(int position) {
        if (position == 0) {
            return mDatas.get(mDatas.size() - 1);
        } else if (position == getCount() - 1) {
            return mDatas.get(0);
        }
        return mDatas.get(position - 1);
    }

    @Override
    public void onPageSelected(int position) {
        int count = getCount();
        if (count != 0) {
            if (position == 0) {
                viewPager.setCurrentItem(count - 2, false);
            } else if (position == count - 1) {
                viewPager.setCurrentItem(1, false);
            }
        }
    }

    @Override
    public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {
    }

    @Override
    public void onPageScrollStateChanged(int state) {

    }
}

File: demo/src/main/java/mocaris/com/demo/LoopViewPagerAdapter2.kt
package mocaris.com.demo

import android.content.Context
import android.support.v4.view.PagerAdapter
import android.support.v4.view.ViewPager

abstract class LoopViewPagerAdapter2<T>(val context: Context, val viewPager: ViewPager) : PagerAdapter(), ViewPager.OnPageChangeListener {
    override fun onPageScrollStateChanged(state: Int) {
    }

    override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {
    }

    override fun onPageSelected(position: Int) {
        if (count != 0) {
            if (position == 0) {
                viewPager.setCurrentItem(count - 2, false)
            } else if (position == count - 1) {
                viewPager.setCurrentItem(1, false)
            }
        }
    }

    private val proverList: ArrayList<T> = ArrayList<T>()
    var mDatas: List<T> = ArrayList<T>()
        set(value) {
            field = value
            if (value.isNotEmpty()) {
                proverList.clear();
                proverList.addAll(field)
                proverList.add(proverList.lastIndex, value.first())//添加第一个
                proverList.add(0, value.last())//添加最后一个
            } else {
                proverList.clear()
            }
            notifyDataSetChanged()
        }

    init {
        viewPager.addOnPageChangeListener(this@LoopViewPagerAdapter2)
    }

    override fun getCount(): Int {
        return proverList?.size ?: 0
    }

    fun getItem(position: Int): T {
        if (position == 0) {
            return mDatas.last()
        } else if (position == count - 1) {
            return mDatas.first()
        }
        return mDatas[position - 1]
    }

    /**
     * 无限轮播
     * 从第一屏开始
     */
}